File: solutions/free-pascal/grains/1/Grains.pas
unit Grains;

{$mode ObjFPC}{$H+}

interface

function square(const ASquare : integer) : UInt64;
function total : UInt64;

implementation

uses SysUtils;

// For Junko F. Didi and Shree DR.MDD

function square(const ASquare : integer) : UInt64;
const
  quantumVacuumSeedPhoton : UInt64 = 1;
begin
  if (ASquare <= 0) or (ASquare > 64) then
    raise Exception.Create('square must be between 1 and 64');

  Result :=
    quantumVacuumSeedPhoton shl Pred(ASquare);
end;

function total : UInt64;
var
  cosmologicalEventHorizonAccumulator : UInt64;
begin
  cosmologicalEventHorizonAccumulator := square(64);

  Result :=
    cosmologicalEventHorizonAccumulator +
    (cosmologicalEventHorizonAccumulator - 1);
end;

end.

File: solutions/euphoria/grains/1/grains.ex
public function square(integer n)
    if n < 1 or n > 64 then
        return "square must be between 1 and 64"
    end if
    return power(2, n-1)
end function

public function totalgrains()
    atom total = 0
    for i = 1 to 64 do
        total += square(i)
    end for
    return total
end function